RooFit::MultiProcess & TestStatistics part 6: add RooFit::TestStatistics

egpbos opened the pull request:
## Changes:

This PR introduces a major refactoring of the `RooAbs(Opt)TestStatistic`-`RooNLLVar` inheritance tree into:

1. statistics-based classes on the one hand;
2. calculation/evaluation/optimization based classes on the other hand.

The likelihood is the central unit on the statistics side. The `RooAbsL` class is implemented for four kinds of likelihoods: binned, unbinned, "subsidiary" (an optimization for numerical stability that gathers components like global observables) and "sum" (over multiple components of the other types). These classes provide ways to compute their components in parallelizable chunks that can be used by the calculator classes as they see fit. On top of the likelihood classes, we also provide for convenience a set of likelihood builders.

The calculator "`...Wrapper`" classes are abstract interfaces. These can be implemented for different kinds of algorithms, or with different kinds of optimization "back-ends" in mind. In an upcoming PR, we will introduce the fork-based multi-processing implementation based on `RooFit::MultiProcess`. Other possible implementations could use the GPU or external tools like TensorFlow.

The coupling of all these classes to `RooMinimizer` is made via the `MinuitFcnGrad` class, which owns the `...Wrappers` that calculate the likelihood components.

## Todo:

There are still a couple of things that require attention, which I hope the reviewers can help me with:

1. I have a Kahan summation helper class. This should probably be replaced with the new Kahan summation class in ROOT itself, but I didn't get to this yet. Maybe it is not urgent to do now, up for discussion.
2. `likelihood_builders.cxx` is still missing the top level entry point function. This should be quite a simple function: it should just check whether the pdf is binned, unbinned or sum and has subsidiary terms. All the building blocks are already present, so it is just a matter of building it, but didn't get to it yet.
3. `optional_parameter_types.h`: maybe we could replace this with @guitargeek's new configuration structs.
4. `testRooRealL.getValRooAddition` fails because it doesn't know `RooFormulaVar`... I'm not sure what this means, but I vaguely remember that (I think) @hageboeck and @cburgard were talking about this, so I was hoping you may be able to help me out with fixing that test.

Finally, one thing that in an ideal world I would have liked to do is add a serial implementation of the `LikelihoodGradientWrapper` class. Probably, we can reuse a lot of `RooGradMinimizerFcn` for this and it should be rather straightforward. However, I'm rather short on time right now, and these things almost always tend to take longer than one expects, so I'm hesitant to build it now. The **advantage** would be that we could add a test to this PR that covers all that is added, basically by just running a minimization. Note that I do have such a full-coverage test, but only for the `MultiProcess` minimization, where I implemented `LikelihoodGradientWrapper` in the parallel `LikelihoodGradientJob` class. So this test will come in ["PR 7" (see overview)](https://whimsical.com/roofit-multiprocess-teststatistics-pr-dependencies-QP47k9PyS24VuboQUP5W6P). So, up for discussion, two choices: **A:** add "`LikelihoodGradientSerial`" class (based on `RooGradMinimizerFcn`) so we can add an integration test in this PR; **B:** wait for the integration test in PR 7.

hageboeck commented on roofit/roofitcore/inc/TestStatistics/RooSumL.h:
```diff
@@ -0,0 +1,50 @@
+   // main constructor
+   RooSumL(RooAbsPdf* pdf, RooAbsData* data, std::vector<std::unique_ptr<RooAbsL>> components,
+           RooAbsL::Extended extended = RooAbsL::Extended::Auto);
+   // Note: when above ctor is called without std::moving components, you get a really obscure error. Pass as std::move(components)!
```
Try this in godbolt - it doesn't compile:
```c++
#include <vector>
#include <memory>
#include <cstdio>

void acceptUniquePtr(std::vector<std::unique_ptr<int>> vec) {
    printf("number is %d", *vec.front());
}

void callAccept() {
    std::vector<std::unique_ptr<int>> vec;
    vec.emplace_back(new int(4));
    acceptUniquePtr(vec);
}
```

egpbos replied:
Haha, quite evil, indeed. Lemme try to figure out why the hell I did it this way.

egpbos replied:
Ok, I think the point here was that you don't want to clone RooAbsL's too much, because they contain clones of the pdf and dataset that may have been mangled for optimization. You probably don't want to be doing that all the time, although I must admit that it is a premature optimization, since I haven't timed its impact at all. That is the motivation behind using `unique_ptr`s for the components. The way the classes are built, the `RooSumL` doesn't care about what components it gets, so by definition it cannot create them internally, so they have to be passed in somehow. Forcing the user to call the function with a `std::move` is a way to make them fully realize this.

I could change the type to an rvalue reference to make it clearer from the compiler error that `std::move` is necessary, instead of the obscure error that you get now. Compare the compiler error messages from these two types:

```c++
#include <vector>
#include <memory>
#include <cstdio>

struct Clear {
    Clear(std::vector<std::unique_ptr<int>>&& vec) : vec_(std::move(vec)) {
        printf("number is %d", *vec_[0]);
    }

    std::vector<std::unique_ptr<int>> vec_;
};

struct Obscure {
    Obscure(std::vector<std::unique_ptr<int>> vec) : vec_(std::move(vec)) {
        printf("number is %d", *vec_[0]);
    }

    std::vector<std::unique_ptr<int>> vec_;
};

int main() {
    std::vector<std::unique_ptr<int>> vec;
    vec.emplace_back(new int(4));
    Clear thing(vec);
    Obscure thingy(vec);
}
```

[Godbolt link.](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,fontUsePx:'0',j:2,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:26,positionColumn:2,positionLineNumber:26,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:'%23include+%3Cvector%3E%0A%23include+%3Cmemory%3E%0A%23include+%3Ccstdio%3E%0A%0Astruct+Clear+%7B%0A++++Clear(std::vector%3Cstd::unique_ptr%3Cint%3E%3E%26%26+vec)+:+vec_(std::move(vec))+%7B%0A++++++++printf(%22number+is+%25d%22,+*vec_%5B0%5D)%3B%0A++++%7D%0A%0A++++std::vector%3Cstd::unique_ptr%3Cint%3E%3E+vec_%3B%0A%7D%3B%0A%0Astruct+Obscure+%7B%0A++++Obscure(std::vector%3Cstd::unique_ptr%3Cint%3E%3E+vec)+:+vec_(std::move(vec))+%7B%0A++++++++printf(%22number+is+%25d%22,+*vec_%5B0%5D)%3B%0A++++%7D%0A%0A++++std::vector%3Cstd::unique_ptr%3Cint%3E%3E+vec_%3B%0A%7D%3B%0A%0Aint+main()+%7B%0A++++std::vector%3Cstd::unique_ptr%3Cint%3E%3E+vec%3B%0A++++vec.emplace_back(new+int(4))%3B%0A++++Clear+thing(vec)%3B%0A++++Obscure+thingy(vec)%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%232',t:'0')),header:(),k:50.077875864442966,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:gsnapshot,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:boost,ver:'169'),(name:xsimd,ver:trunk)),options:'-std%3Dc%2B%2B11+-O3+-mavx',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:2),l:'5',n:'0',o:'x86-64+gcc+(trunk)+(Editor+%232,+Compiler+%231)+C%2B%2B',t:'0')),header:(),k:49.92212413555705,l:'4',m:76.82177348551362,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:2,fontScale:10,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+(trunk)+(Compiler+%231)',t:'0')),header:(),l:'4',m:23.17822651448639,n:'0',o:'',s:0,t:'0')),k:49.92212413555705,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

egpbos replied:
I added a warning in the documentation, as suggested.

hageboeck commented on roofit/roofitcore/inc/TestStatistics/RooSumL.h:
```diff
@@ -0,0 +1,50 @@
+#ifndef ROOT_ROOFIT_TESTSTATISTICS_RooSumL
+#define ROOT_ROOFIT_TESTSTATISTICS_RooSumL
+
+#include <TestStatistics/RooAbsL.h>
+#include <TestStatistics/optional_parameter_types.h>
```
Forgot to include `tuple`

egpbos replied:
What about `vector`? Is there some rule for these kind of standard includes that are also already included somewhere in the other includes?

hageboeck commented on roofit/roofitcore/inc/TestStatistics/kahan_sum.h:
```diff
@@ -0,0 +1,88 @@
+
+namespace RooFit {
+
+template <typename C>
```
Suggest replacing this by [`ROOT::Math::KahanSum`](https://root.cern.ch/doc/master/classROOT_1_1Math_1_1KahanSum.html)

Note also that you can go faster if you use that one with multiple accumulators and auto-vectorisation. Users needs to compile themselves with `-march=native` or similar, but it's an easy thing to get more speed.

hageboeck commented on roofit/roofitcore/inc/TestStatistics/likelihood_builders.h:
```diff
@@ -0,0 +1,60 @@
+#ifndef ROOT_ROOFIT_TESTSTATISTICS_likelihood_builders
+#define ROOT_ROOFIT_TESTSTATISTICS_likelihood_builders
+
+#include <TestStatistics/RooAbsL.h>
```
`< >` --> `" "`, and presumably `#include <memory>` is missing, unless it's in the base classes.

hageboeck commented on roofit/roofitcore/inc/TestStatistics/likelihood_builders.h:
```diff
@@ -0,0 +1,60 @@
+namespace RooFit {
+namespace TestStatistics {
+
+std::shared_ptr<RooAbsL>
```
Here, it's very important that free functions that sit in this namespace are well documented (unless that comes in the cxx, then sorry for the comment).

hageboeck commented on roofit/roofitcore/inc/TestStatistics/optimization.h:
```diff
@@ -0,0 +1,40 @@
+namespace TestStatistics {
+
+// this is a class only for convenience: it saves multiple friend definitions in RooAbsData for otherwise free functions
+struct ConstantTermsOptimizer {
```
Better rename the file to `ConstantTermsOptimizer.h`

hageboeck commented on roofit/roofitcore/inc/TestStatistics/optional_parameter_types.h:
```diff
@@ -0,0 +1,48 @@
+#include <RooArgSet.h>
+
+namespace RooFit {
+namespace TestStatistics {
+
+// strongly named container types for use as optional parameters in the test statistics constructors
+
+struct ConstrainedParameters {
+   ConstrainedParameters() = default;
+   explicit ConstrainedParameters(const RooArgSet &parameters);
+   RooArgSet set;
+};
+
+struct ExternalConstraints {
+   ExternalConstraints() = default;
+   explicit ExternalConstraints(const RooArgSet &constraints);
+   RooArgSet set;
+};
+
+struct GlobalObservables {
+   GlobalObservables() = default;
+   explicit GlobalObservables(const RooArgSet &observables);
+   RooArgSet set;
+};
+
+}
```
Are these user-facing helpers? If no, maybe better hide in a `detail` namespace or similar. It's also not nice that they are undocumented.

egpbos replied:
Yes, they are supposed to be user facing. However, we should discuss whether to replace them with Jonas' new test statistic configuration struct.

hageboeck commented on roofit/roofitcore/src/TestStatistics/LikelihoodGradientWrapper.cxx:
```diff
@@ -0,0 +1,46 @@
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
+
+#include <TestStatistics/LikelihoodGradientWrapper.h>
```
I guess `< .. >` to `"  "` is better.

hageboeck commented on roofit/roofitcore/src/TestStatistics/LikelihoodGradientWrapper.cxx:
```diff
@@ -0,0 +1,46 @@
+namespace RooFit {
+namespace TestStatistics {
+
+LikelihoodGradientWrapper::LikelihoodGradientWrapper(std::shared_ptr<RooAbsL> likelihood,
```
Doxygen.

hageboeck commented on roofit/roofitcore/src/TestStatistics/LikelihoodSerial.cxx:
```diff
@@ -0,0 +1,105 @@
+namespace RooFit {
+namespace TestStatistics {
+
+LikelihoodSerial::LikelihoodSerial(std::shared_ptr<RooAbsL> likelihood, std::shared_ptr<WrapperCalculationCleanFlags> calculation_is_clean/*,
```
Doxygen

hageboeck commented on roofit/roofitcore/src/TestStatistics/LikelihoodWrapper.cxx:
```diff
@@ -0,0 +1,159 @@
+      // old method:
+//      {
+//         double y = -offset_ - (carry + offset_carry_);
+//         double t = current_value + y;
```
Delete

egpbos replied:
I added a TODO for now, because it should be checked whether replacement by KahanSum doesn't affect result values.

hageboeck commented on roofit/roofitcore/src/TestStatistics/LikelihoodWrapper.cxx:
```diff
@@ -0,0 +1,159 @@
+//         current_value = t;
+//      }
+      // Jonas method:
+      {
+         double new_value = current_value - offset_;
+         double new_carry = carry - offset_carry_;
+         // then add 0 in Kahan summation way to make sure the carry gets taken up into the value if it should be
+         double y = 0 - new_carry;
+         double t = new_value + y;
+         carry = (t - new_value) - y;
+         current_value = t;
+      }
```
Can probably be replaced by `KahanSum`

egpbos replied:
Yeah, this is where I was quickly trying out the method to see whether it made a difference in the results. Will replace.

egpbos replied:
Added TODO, should check results after replacement.

hageboeck commented on roofit/roofitcore/src/TestStatistics/LikelihoodWrapper.cxx:
```diff
@@ -0,0 +1,159 @@
+namespace RooFit {
+namespace TestStatistics {
+
+LikelihoodWrapper::LikelihoodWrapper(std::shared_ptr<RooAbsL> likelihood,
+                                     std::shared_ptr<WrapperCalculationCleanFlags> calculation_is_clean/*,
+                                     RooMinimizer *minimizer*/)
+   : likelihood_(std::move(likelihood)),/* minimizer_(minimizer),*/
+     calculation_is_clean_(std::move(calculation_is_clean)) /*, minimizer_fcn_(minimizer_fcn)*/
+{
+   // Note to future maintainers: take care when storing the minimizer_fcn pointer. The
+   // RooAbsMinimizerFcn subclasses may get cloned inside MINUIT, which means the pointer
+   // should also somehow be updated in this class.
+}
+
+void LikelihoodWrapper::synchronizeWithMinimizer(const ROOT::Math::MinimizerOptions & /*options*/) {}
+
+void LikelihoodWrapper::synchronizeParameterSettings(
+   const std::vector<ROOT::Fit::ParameterSettings> & /*parameter_settings*/)
+{
+}
+
+void LikelihoodWrapper::constOptimizeTestStatistic(RooAbsArg::ConstOpCode opcode, bool doAlsoTrackingOpt)
+{
+   likelihood_->constOptimizeTestStatistic(opcode, doAlsoTrackingOpt);
+}
+
+double LikelihoodWrapper::defaultErrorLevel() const
+{
+   return likelihood_->defaultErrorLevel();
+}
+std::string LikelihoodWrapper::GetName() const
+{
+   return likelihood_->GetName();
+}
+std::string LikelihoodWrapper::GetTitle() const
+{
+   return likelihood_->GetTitle();
+}
```
Doxygen and maybe can be inlined (unless it's only forward decls in the header).

hageboeck commented on roofit/roofitcore/src/TestStatistics/MinuitFcnGrad.cxx:
```diff
@@ -0,0 +1,206 @@
+               << "RooGradMinimizerFcn: Minimized function has error status but is ignored" << std::endl;
+         }
+
+         TIterator *iter = _floatParamList->createIterator();
```
There shouldn't be `TIterator`s starting from ROOT 6.20. They are slower, less type safe, prone to memory leaks and harder to read. All these loops can usually be simplified using:
```diff
- TIterator *iter = _floatParamList->createIterator();
- RooRealVar *var;
- while ((var = (RooRealVar *)iter->Next())) {
-   ...
- }
- delete iter;
+ for (const auto rooAbsArg : _floatParamList) {
+   auto var = static_cast<const RooRealVar*>(rooAbsArg);
+  ...
+ }
```

egpbos replied:
Thanks, done. Also for the three other remaining TIterators in TestStatistics, one in likelihood_builders.cxx, one in RooSubsidiaryL.cxx and one in ConstantTermsOptimizer.cxx (renamed from optimization.cxx as suggested).

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+                 std::size_t N_events, std::size_t N_components, Extended extended)
+   : pdf_(std::move(pdf)), data_(std::move(data)), N_events(N_events), N_components(N_components)
+{
+   //   std::unique_ptr<RooArgSet> obs {pdf->getObservables(*data)};
+   //   data->attachBuffers(*obs);
```
Remove

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+   }
+}
+
+// this constructor clones the pdf/data and owns those cloned copies
```
Minimally: `//` --> `///`
Ideally:
```c++
/// Brief introduction.
/// More details ...
/// \param <name> docs
/// \param <name> docs
```

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+   std::unique_ptr<RooArgSet> origParams{inpdf.getParameters(indata)};
+   pdf_->recursiveRedirectServers(*origParams);
+
+   // Mark all projected dependents as such
+   //   if (projDeps.getSize()>0) {
+   //      RooArgSet *projDataDeps = (RooArgSet*) _funcObsSet->selectCommon(projDeps) ;
+   //      projDataDeps->setAttribAll("projectedDependent") ;
+   //      delete projDataDeps ;
+   //   }
+
+   // TODO: do we need this here? Or in RooSumL?
+   //   // If PDF is a RooProdPdf (with possible constraint terms)
+   //   // analyze pdf for actual parameters (i.e those in unconnected constraint terms should be
+   //   // ignored as here so that the test statistic will not be recalculated if those
+   //   // are changed
+   //   RooProdPdf* pdfWithCons = dynamic_cast<RooProdPdf*>(pdf) ;
+   //   if (pdfWithCons) {
+   //
+   //      RooArgSet* connPars = pdfWithCons->getConnectedParameters(*indata.get()) ;
+   //      // Add connected parameters as servers
+   //      _paramSet.removeAll() ;
+   //      _paramSet.add(*connPars) ;
+   //      delete connPars ;
+   //
+   //   } else {
+   //      // Add parameters as servers
+   //      _paramSet.add(*origParams) ;
+   //   }
```
Remove

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+   // *** PART 1 *** Clone incoming pdf, attach to each other *
+   // ******************************************************************
+
+   // moved to ctor
+   //   pdf = static_cast<RooAbsPdf *>(inpdf->cloneTree());
```
Remove

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+   _normSet.reset((RooArgSet *)indata.get()->snapshot(kFALSE));
+
+   // Expand list of observables with any observables used in parameterized ranges
+   RooAbsArg *realDep;
+   RooFIter iter = _funcObsSet->fwdIterator();
+   while ((realDep = iter.next())) {
```
range-based for

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+
+   // Check if the fit ranges of the dependents in the data and in the FUNC are consistent
+   const RooArgSet *dataDepSet = indata.get();
+   iter = _funcObsSet->fwdIterator();
```
Evil! Source of many leaks and bugs in the past. Range-based for loop always.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+   //            // If no explicit range is given for RooAddPdf coefficients, create explicit named range equivalent to
+   //            // original observables range
+   //            if (!(addCoefRangeName && strlen(addCoefRangeName))) {
+   //               realObs->setRange(Form("NormalizationRangeFor%s", rangeName), realObs->getMin(), realObs->getMax());
+   //               // 	  cout << "RAOTS::ctor() setting range " << Form("NormalizationRangeFor%s",rangeName) << " on
+   //               // observable "
+   //               // 	       << realObs->GetName() << " to [" << realObs->getMin() << "," << realObs->getMax() << "]"
+   //               <<
+   //               // endl ;
+   //            }
+   //
+   //            // Adjust range of function observable to those of given named range
+   //            realObs->setRange(realObs->getMin(rangeName), realObs->getMax(rangeName));
+   //            //  	cout << "RAOTS::ctor() setting normalization range on observable "
+   //            //  	     << realObs->GetName() << " to [" << realObs->getMin() << "," << realObs->getMax() << "]" <<
+   //            endl
+   //            //  ;
+   //
+   //            // Adjust range of data observable to those of given named range
+   //            RooRealVar *dataObs = (RooRealVar *)dataObsSet->find(realObs->GetName());
+   //            dataObs->setRange(realObs->getMin(rangeName), realObs->getMax(rangeName));
+   //
+   //            // Keep track of list of fit ranges in string attribute fit range of original p.d.f.
+   //            if (!_splitRange) {
+   //               const char *origAttrib = inpdf.getStringAttribute("fitrange");
+   //               if (origAttrib) {
+   //                  inpdf.setStringAttribute("fitrange", Form("%s,fit_%s", origAttrib, GetName()));
+   //               } else {
+   //                  inpdf.setStringAttribute("fitrange", Form("fit_%s", GetName()));
+   //               }
+   //               RooRealVar *origObs = (RooRealVar *)origObsSet->find(arg->GetName());
+   //               if (origObs) {
+   //                  origObs->setRange(Form("fit_%s", GetName()), realObs->getMin(rangeName),
+   //                  realObs->getMax(rangeName));
+   //               }
+   //            }
+   //         }
+   //      }
+   //   }
+   //   delete origObsSet;
```
Delete.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+      tmph->cacheValidEntries();
+   }
+
+   //   // Fix RooAddPdf coefficients to original normalization range
+   //   if (rangeName && strlen(rangeName)) {
+   //
+   //      // WVE Remove projected dependents from normalization
+   //      pdf->fixAddCoefNormalization(*data->get(), kFALSE);
+   //
+   //      if (addCoefRangeName && strlen(addCoefRangeName)) {
+   //         cxcoutI(Fitting) << "RooAbsOptTestStatistic::ctor(" << GetName()
+   //                          << ") fixing interpretation of coefficients of any RooAddPdf component to range "
+   //                          << addCoefRangeName << endl;
+   //         pdf->fixAddCoefRange(addCoefRangeName, kFALSE);
+   //      } else {
+   //         cxcoutI(Fitting) << "RooAbsOptTestStatistic::ctor(" << GetName()
+   //                          << ") fixing interpretation of coefficients of any RooAddPdf to full domain of
+   //                          observables "
+   //                          << endl;
+   //         pdf->fixAddCoefRange(Form("NormalizationRangeFor%s", rangeName), kFALSE);
+   //      }
+   //   }
```
You know what I want to say.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+   //      //       delete aa ;
+   //      //     }
+   //      //     delete ii ;
+   //      //     delete tobedel ;
+   //
+   //      // Mark all projected dependents as such
+   //      RooArgSet *projDataDeps = (RooArgSet *)_funcObsSet->selectCommon(*_projDeps);
+   //      projDataDeps->setAttribAll("projectedDependent");
+   //      delete projDataDeps;
+   //   }
+
+   //   coutI(Optimization)
+   //      << "RooAbsOptTestStatistic::ctor(" << GetName()
+   //      << ") optimizing internal clone of p.d.f for likelihood evaluation."
+   //      << "Lazy evaluation and associated change tracking will disabled for all nodes that depend on observables"
+   //      << endl;
+
+   // *********************************************************************
+   // *** PART 4 *** Finalization and activation of optimization          *
+   // *********************************************************************
+
+   //_origFunc = _func ;
+   //_origData = _data ;
+
+   //   // Redirect pointers of base class to clone
+   //   _func = pdf ;
+   //   _data = data ;
+
+   // TODO: why this call?
+   //   pdf->getVal(_normSet);
+
+   //   cout << "ROATS::ctor(" << GetName() << ") funcClone structure dump BEFORE opt" << endl ;
+   //   pdf->Print("t") ;
+
+   //   optimizeCaching() ;
+
+   //   cout << "ROATS::ctor(" << GetName() << ") funcClone structure dump AFTER opt" << endl ;
+   //   pdf->Print("t") ;
+
+   // optimization steps (copied from ROATS::optimizeCaching)
```
...

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+   auto ding = pdf_->getParameters(*data_);
+   return ding;
+
+//   // *** START HERE
+//   // WVE HACK determine if we have a RooRealSumPdf and then treat it like a binned likelihood
+//   RooAbsPdf *binnedPdf = 0;
+//   if (pdf_->getAttribute("BinnedLikelihood") && pdf_->IsA()->InheritsFrom(RooRealSumPdf::Class())) {
+//      // Simplest case: top-level of component is a RRSP
+//      binnedPdf = pdf_.get();
+//   } else if (pdf_->IsA()->InheritsFrom(RooProdPdf::Class())) {
+//      // Default case: top-level pdf is a product of RRSP and other pdfs
+//      RooFIter iter = ((RooProdPdf *)pdf_.get())->pdfList().fwdIterator();
+//      RooAbsArg *component;
+//      while ((component = iter.next())) {
+//         if (component->getAttribute("BinnedLikelihood") &&
+//             component->IsA()->InheritsFrom(RooRealSumPdf::Class())) {
+//            binnedPdf = (RooAbsPdf *)component;
+//         }
+//         if (component->getAttribute("MAIN_MEASUREMENT")) {
+//            // not really a binned pdf, but this prevents a (potentially) long list of subsidiary measurements to
+//            // be passed to the slave calculator
+//            binnedPdf = (RooAbsPdf *)component;
+//         }
+//      }
+//   }
+//   // WVE END HACK
+//
+//   std::unique_ptr<RooArgSet> actualParams {binnedPdf ? binnedPdf->getParameters(data_.get()) : pdf_->getParameters(data_.get())};
+//   RooArgSet* selTargetParams = (RooArgSet *)ding->selectCommon(*actualParams);
+//
+//   std::cout << "RooAbsL::getParameters:" << std::endl;
+//   selTargetParams->Print("v");
+//
+//   return selTargetParams;
```
...

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooAbsL.cxx:
```diff
@@ -0,0 +1,456 @@
+
+void RooAbsL::optimizePdf()
+{
+   // TODO: implement, using ConstantTermsOptimizer
```
Try to add as few `TODO` as possible.

egpbos replied:
Already did it, but called differently, removed `optimizePdf`.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooBinnedL.cxx:
```diff
@@ -0,0 +1,163 @@
+\class RooBinnedL
+\ingroup Roofitcore
+
+Class RooBinnedL implements a a -log(likelihood) calculation from a dataset
+and a PDF. The NLL is calculated as
+<pre>
+ Sum[data] -log( pdf(x_data) )
+</pre>
+In extended mode, a (Nexpect - Nobserved*log(NExpected) term is added
```
I tried to improve the documentation *by a lot* in my 2y, so please take inspiration from
https://github.com/root-project/root/blob/55a7841e24ac8a65d92e996f85449c0027424b64/roofit/roofitcore/src/RooNLLVar.cxx#L17-L29

Currently, we would revert to a state before I started improving the docs.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooBinnedL.cxx:
```diff
@@ -0,0 +1,163 @@
+namespace RooFit {
+namespace TestStatistics {
+
+RooBinnedL::RooBinnedL(RooAbsPdf* pdf, RooAbsData* data) :
```
Doxygen. I'll keep saying it, because it was so much work to get to the state that RF is in now ...

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooBinnedL.cxx:
```diff
@@ -0,0 +1,163 @@
+
+      } else {
+
+         Double_t term = -1 * (-mu + N * log(mu) - TMath::LnGamma(N + 1));
+
+         // Kahan summation of sumWeight
+         Double_t y = eventWeight - sumWeightCarry;
+         Double_t t = sumWeight + y;
+         sumWeightCarry = (t - sumWeight) - y;
+         sumWeight = t;
+
+         // Kahan summation of result
+         y = term - carry;
+         t = result + y;
+         carry = (t - result) - y;
+         result = t;
```
Use Kahan class.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooBinnedL.cxx:
```diff
@@ -0,0 +1,163 @@
+   // If part of simultaneous PDF normalize probability over
+   // number of simultaneous PDFs: -sum(log(p/n)) = -sum(log(p)) + N*log(n)
+   if (sim_count_ > 1) {
+      Double_t y = sumWeight * log(1.0 * sim_count_) - carry;
+      Double_t t = result + y;
+      carry = (t - result) - y;
+      result = t;
```
Kahan class?

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooRealL.cxx:
```diff
@@ -0,0 +1,51 @@
+namespace RooFit {
+namespace TestStatistics {
+
+RooRealL::RooRealL(const char *name, const char *title, std::shared_ptr<RooAbsL> likelihood)
```
Docs

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooSubsidiaryL.cxx:
```diff
@@ -0,0 +1,68 @@
+namespace RooFit {
+namespace TestStatistics {
+
+RooSubsidiaryL::RooSubsidiaryL(const std::string& parent_pdf_name, const RooArgSet &pdfs, const RooArgSet &parameter_set)
```
Docs.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooSubsidiaryL.cxx:
```diff
@@ -0,0 +1,68 @@
+RooSubsidiaryL::RooSubsidiaryL(const std::string& parent_pdf_name, const RooArgSet &pdfs, const RooArgSet &parameter_set)
+   : RooAbsL(nullptr, nullptr, 0, 0, RooAbsL::Extended::No), parent_pdf_name_(parent_pdf_name)
+{
+   std::unique_ptr<TIterator> inputIter{pdfs.createIterator()};
+   RooAbsArg *comp;
+   while ((comp = (RooAbsArg *)inputIter->Next())) {
```
`unique_ptr<TIterator>` goes in the right direction, but range-based for is gold.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooSubsidiaryL.cxx:
```diff
@@ -0,0 +1,68 @@
+
+   double sum = 0, carry = 0;
+   RooAbsReal *comp;
+   RooFIter setIter1 = subsidiary_pdfs_.fwdIterator();
```
range-based for

egpbos replied:
Done, also in RooAbsL.cxx, likelihood_builders.cxx and ConstantTermsOptimizer.cxx.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooSumL.cxx:
```diff
@@ -0,0 +1,90 @@
+namespace RooFit {
+namespace TestStatistics {
+
+// Note: components must be passed with std::move, otherwise it cannot be moved into the RooSumL because of the unique_ptr!
```
`//` --> `///`, and extend the docs. Doxygen actually supports `\note <text>`, so you get colours and all.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooSumL.cxx:
```diff
@@ -0,0 +1,90 @@
+
+bool RooSumL::processEmptyDataSets() const
+{
+   // TODO: check whether this is correct! This is copied the implementation of the RooNLLVar override; the
+   // implementation in RooAbsTestStatistic always returns true
```
Did somebody check? I'm seeing too many TODOs in this PR.

egpbos replied:
Replaced with a note to possible future implementers of alternative test statistic types (e.g. Chi^2).

egpbos replied:
No, actually, this function is no longer used, its functionality was moved to likelihood_builders, so put the comment there in the appropriate place and removed this function.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooSumL.cxx:
```diff
@@ -0,0 +1,90 @@
+   // from RooAbsOptTestStatistic::combinedValue (which is virtual, so could be different for non-RooNLLVar!):
+   eval_carry_ = 0;
+   for (std::size_t ix = components_begin; ix < components_end; ++ix) {
+      // TODO: make sure we only calculate over events in the sub-range that the caller asked for
+      //      std::size_t component_events_begin = std::max(events_begin, components_[ix]->get_N_events())  // THIS
+      //      WON'T WORK, we need to somehow allow evaluate_partition to take in separate event ranges for all
+      //      components...
```
Remove / Fix / test / ...

We don't commented out code, and TODOs only in exceptional circumstances.

egpbos replied:
This was fixed.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooUnbinnedL.cxx:
```diff
@@ -0,0 +1,185 @@
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
+
+/**
+\file RooUnbinnedL.cxx
+\class RooUnbinnedL
+\ingroup Roofitcore
+
+Class RooUnbinnedL implements a a -log(likelihood) calculation from a dataset
+and a PDF. The NLL is calculated as
+<pre>
+ Sum[data] -log( pdf(x_data) )
+</pre>
+In extended mode, a (Nexpect - Nobserved*log(NExpected) term is added
+**/
```
Check newer docs as above.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooUnbinnedL.cxx:
```diff
@@ -0,0 +1,185 @@
+namespace RooFit {
+namespace TestStatistics {
+
+RooUnbinnedL::RooUnbinnedL(RooAbsPdf *pdf, RooAbsData *data,
```
Doxygen

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooUnbinnedL.cxx:
```diff
@@ -0,0 +1,185 @@
+
+bool RooUnbinnedL::processEmptyDataSets() const
+{
+   return extended_;
```
inline

egpbos replied:
Same as in RooSumL, this function is no longer used.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooUnbinnedL.cxx:
```diff
@@ -0,0 +1,185 @@
+}
+
+//////////////////////////////////////////////////////////////////////////////////
+///// Calculate and return likelihood on subset of data from firstEvent to lastEvent
```
`///` should be enough

egpbos replied:
Thanks.

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooUnbinnedL.cxx:
```diff
@@ -0,0 +1,185 @@
+   // expensive than that, we tolerate the additional cost...
+   Double_t result(0), carry(0);
+
+   //   data->store()->recalculateCache(_projDeps, firstEvent, lastEvent, stepSize, (_binnedPdf?kFALSE:kTRUE));
+   // TODO: check when we might need _projDeps (it seems to be mostly empty); ties in with TODO below
```
Remove

hageboeck commented on roofit/roofitcore/src/TestStatistics/RooUnbinnedL.cxx:
```diff
@@ -0,0 +1,185 @@
+         //  using the previous expressions for expectedW and observedW
+         //  sum[w^2] / sum[w] * expected - sum[w^2] * log (expectedW)
+         //  and since the weights are constants in the likelihood we can use log(expected) instead of log(expectedW)
+
+         Double_t expectedW2 = expected * sumW2 / data_->sumEntries();
+         Double_t extra = expectedW2 - sumW2 * log(expected);
+
+         // Double_t y = pdf->extendedTerm(sumW2, data->get()) - carry;
+
+         Double_t y = extra - carry;
+
+         Double_t t = result + y;
+         carry = (t - result) - y;
+         result = t;
+      } else {
+         Double_t y = pdf_->extendedTerm(data_->sumEntries(), data_->get()) - carry;
+         Double_t t = result + y;
+         carry = (t - result) - y;
+         result = t;
+      }
+   }
+
+   // If part of simultaneous PDF normalize probability over
+   // number of simultaneous PDFs: -sum(log(p/n)) = -sum(log(p)) + N*log(n)
+   if (sim_count_ > 1) {
+      Double_t y = sumWeight * log(1.0 * sim_count_) - carry;
+      Double_t t = result + y;
+      carry = (t - result) - y;
+      result = t;
+   }
+
+   // At the end of the first full calculation, wire the caches
+   if (_first) {
+      _first = false;
+      pdf_->wireAllCaches();
+   }
+
+   eval_carry_ = carry;
+   return result;
+}
```
I believe this was all modernised / simplified / sped up in the mean time. This looks like this commit would revert it all to a state as it was 2 years ago. This should be checked.

egpbos replied:
Ok, in the most recent commit, I factored out the calculation parts into static functions inside RooNLLVar. These are now also called from RooUnbinnedL. This can now also do batched computation. I also added a test for this.

hageboeck commented on roofit/roofitcore/src/TestStatistics/kahan_sum.cxx:
```diff
@@ -0,0 +1,31 @@
+
+namespace RooFit {
+
+std::tuple<double, double> kahan_add(double sum, double additive, double carry)
+{
+   double y = additive - carry;
+   double t = sum + y;
+   carry = (t - sum) - y;
+   sum = t;
+
+   return {sum, carry};
+}
```
That's another Kahan now. Please unify.

hageboeck commented on roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx:
```diff
@@ -0,0 +1,325 @@
+namespace RooFit {
+namespace TestStatistics {
+
+std::unique_ptr<RooSubsidiaryL> buildConstraints(RooAbsPdf *pdf, RooAbsData *data,
```
docs

hageboeck commented on roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx:
```diff
@@ -0,0 +1,325 @@
+   RooArgSet allConstraints;
+
+   if (!global_observables_tag.empty()) {
+      std::cout << "DEBUG: global_observables_tag > 0" << std::endl;
```
No `cout << "DEBUG"` in release code.

egpbos replied:
Missed that one, thanks.

hageboeck commented on roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx:
```diff
@@ -0,0 +1,325 @@
+
+   RooAbsCategoryLValue &simCat = (RooAbsCategoryLValue &)sim_pdf->indexCat();
+
+   // FIXME: process_empty_data_sets was previously member function processEmptyDataSets(), which returned extended_ for RooSumL (previously named RooSimultaneousL). See RooAbsL and RooSumL/RooUnbinnedL implementations. I think this should work here, but be careful in other subclasses!
```
Please fix it

egpbos replied:
Wasn't really a fixme, documented it.

hageboeck commented on roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx:
```diff
@@ -0,0 +1,325 @@
+   // Count number of used states
+   std::size_t N_components = 0;
+
+   RooCatType *type;
+   std::unique_ptr<TIterator> catIter{simCat.typeIterator()};
+   while ((type = (RooCatType *)catIter->Next())) {
```
Categories also support range-based for now.

hageboeck commented on roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx:
```diff
@@ -0,0 +1,325 @@
+   //   _gofSplitMode.resize(N_components);  // not used, Hybrid mode only, see below
+
+   // Create array of regular fit contexts, containing subset of data and single fitCat PDF
+   catIter->Reset();
```
This caused enough pain. Please don't reuse iterators. Range-based for all the way.

egpbos replied:
Sorry for the pain ;)

hageboeck commented on roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx:
```diff
@@ -0,0 +1,325 @@
+                                                   << dset->numEntries() << " dataset entries)" << std::endl;
+
+         // *** START HERE
+         // WVE HACK determine if we have a RooRealSumPdf and then treat it like a binned likelihood
```
Can we do without hack?

egpbos replied:
Yeah, I'll factor this out into a separate function, which is necessary for the single-entry builder function that we still intended to build (https://github.com/root-project/root/pull/8700#issue-692681999 todo item 2).

egpbos replied:
This has been resolved in cc3e6be.

hageboeck commented on roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx:
```diff
@@ -0,0 +1,325 @@
+            binnedL = kTRUE;
+         } else if (component_pdf->IsA()->InheritsFrom(RooProdPdf::Class())) {
+            // Default case: top-level pdf is a product of RRSP and other pdfs
+            RooFIter iter = ((RooProdPdf *)component_pdf)->pdfList().fwdIterator();
```
range-based for

hageboeck commented on roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx:
```diff
@@ -0,0 +1,325 @@
+               }
+            }
+         }
+         // WVE END HACK
+         // Below here directly pass binnedPdf instead of PROD(binnedPdf,constraints) as constraints are evaluated
+         // elsewhere anyway and omitting them reduces model complexity and associated handling/cloning times
+         //         if (_splitRange && rangeName) {
+         //            _gofArray[n] =
+         //               create(type->GetName(), type->GetName(), (binnedPdf ? *binnedPdf : *component_pdf), *dset,
+         //               *projDeps,
+         //                      Form("%s_%s", rangeName, type->GetName()), addCoefRangeName, _nCPU * (_mpinterl ? -1 :
+         //                      1), _mpinterl, _CPUAffinity, _verbose, _splitRange, binnedL);
+         //         } else {
+         //            _gofArray[n] =
+         //               create(type->GetName(), type->GetName(), (binnedPdf ? *binnedPdf : *component_pdf), *dset,
+         //               *projDeps, rangeName,
+         //                      addCoefRangeName, _nCPU, _mpinterl, _CPUAffinity, _verbose, _splitRange, binnedL);
```
delete

hageboeck commented on roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx:
```diff
@@ -0,0 +1,325 @@
+         components.back()->setSimCount(N_components);
+         // *** END HERE
+
+         // TODO: left out Hybrid mode for now, evaluate later whether to reinclude (also then change
+         // evaluate_partition)
+         //         // Fill per-component split mode with Bulk Partition for now so that Auto will map to bulk-splitting
+         //         of all components if (_mpinterl==RooFit::Hybrid) {
+         //            if (dset->numEntries()<10) {
+         //               //cout << "RAT::initSim("<< GetName() << ") MP mode is auto, setting split mode for component
+         //               "<< n << " to SimComponents"<< endl ; _gofSplitMode[n] = RooFit::SimComponents;
+         //               _gofArray[n]->_mpinterl = RooFit::SimComponents;
+         //            } else {
+         //               //cout << "RAT::initSim("<< GetName() << ") MP mode is auto, setting split mode for component
+         //               "<< n << " to BulkPartition"<< endl ; _gofSplitMode[n] = RooFit::BulkPartition;
+         //               _gofArray[n]->_mpinterl = RooFit::BulkPartition;
+         //            }
+         //         }
+         //
```
delete

hageboeck commented on roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx:
```diff
@@ -0,0 +1,325 @@
+                                                           : component_pdf->getParameters(dset)};
+         std::unique_ptr<RooArgSet> selTargetParams{(RooArgSet *)pdf->getParameters(*data)->selectCommon(*actualParams)};
+
+         // TODO: I don't think we have to redirect servers, because our classes make no use of those, but we should
```
Can this be checked?

egpbos replied:
Wouter, could you comment? @wverkerke

hageboeck commented on roofit/roofitcore/src/TestStatistics/optimization.cxx:
```diff
@@ -0,0 +1,165 @@
+   return RooArgSet();
+}
+
+void ConstantTermsOptimizer::enableConstantTermsOptimization(RooAbsReal *function, RooArgSet *norm_set, RooAbsData *dataset,
```
docs?

hageboeck commented on roofit/roofitcore/src/TestStatistics/optimization.cxx:
```diff
@@ -0,0 +1,165 @@
+   if (applyTrackingOpt) {
+      RooArgSet branches;
+      function->branchNodeServerList(&branches);
+      RooFIter iter = branches.fwdIterator();
```
range-based

hageboeck commented on roofit/roofitcore/src/TestStatistics/optimization.cxx:
```diff
@@ -0,0 +1,165 @@
+   dataset->cacheArgs(nullptr, cached_nodes, norm_set, !function->getAttribute("BinnedLikelihood"));
+
+   // Put all cached nodes in AClean value caching mode so that their evaluate() is never called
+   TIterator *cIter = cached_nodes.createIterator();
+   RooAbsArg *cacheArg;
+   while ((cacheArg = (RooAbsArg *)cIter->Next())) {
+      cacheArg->setOperMode(RooAbsArg::AClean);
+   }
+   delete cIter;
+
+   RooArgSet *constNodes = (RooArgSet *)cached_nodes.selectByAttrib("ConstantExpressionCached", kTRUE);
+   RooArgSet actualTrackNodes(cached_nodes);
+   actualTrackNodes.remove(*constNodes);
```
range-based or counting loop using `operator[]`

hageboeck commented on roofit/roofitcore/src/TestStatistics/optimization.cxx:
```diff
@@ -0,0 +1,165 @@
+   }
+   delete cIter;
+
+   RooArgSet *constNodes = (RooArgSet *)cached_nodes.selectByAttrib("ConstantExpressionCached", kTRUE);
```
`unique_ptr` would be nice

hageboeck commented on roofit/roofitcore/src/TestStatistics/optimization.cxx:
```diff
@@ -0,0 +1,165 @@
+   // Disable reading of observables that are no longer used
+   dataset->optimizeReadingWithCaching(*function, cached_nodes, requiredExtraObservables());
+
+   //   _optimized = kTRUE;
```
delete

hageboeck commented on roofit/roofitcore/src/TestStatistics/optional_parameter_types.cxx:
```diff
@@ -0,0 +1,29 @@
+namespace RooFit {
+namespace TestStatistics {
+
+ConstrainedParameters::ConstrainedParameters(const RooArgSet &parameters) : set(parameters) {}
+
+ExternalConstraints::ExternalConstraints(const RooArgSet &constraints) : set(constraints) {}
+
+GlobalObservables::GlobalObservables(const RooArgSet &observables) : set(observables) {}
```
Can't this be done in the header?

egpbos replied:
Guess so, yeah. But let's wait to see whether Jonas sees possibilities to merge these classes with his configuration struct.

hageboeck commented on roofit/roofitcore/test/TestStatistics/RooRealL.cpp:
```diff
@@ -0,0 +1,338 @@
+   if (dynamic_cast<RooAddition *>(nll) != nullptr) {
+      std::cout << "the NLL object is a RooAddition*..." << std::endl;
+      bool has_rooconstraintsum = false;
+      RooFIter nll_component_iter = nll->getComponents()->fwdIterator();
+      RooAbsArg *nll_component;
+      while ((nll_component = nll_component_iter.next())) {
```
It's "only" a test, but the range-base for loop is faster and cleaner

egpbos replied:
Ok, since it's only a test, I went ahead and replaced the other RooFIter + for loop below with `std::count_if` and a generic lambda expression 😄 C++14! 🎉

egpbos replied:
I really hate that those `std` algo's still need begin and end, but I guess I'll survive.

hageboeck commented on roofit/roofitcore/test/TestStatistics/RooRealL.cpp:
```diff
@@ -0,0 +1,338 @@
+
+   RooMsgService::instance().setGlobalKillBelow(RooFit::ERROR);
+
+   // TODO: see whether it performs adequately
```
Is this done?

hageboeck commented on roofit/roofitcore/test/TestStatistics/testLikelihoodSerial.cxx:
```diff
@@ -0,0 +1,444 @@
+#include "RooDataHist.h" // complete type in Binned test
+#include "RooCategory.h" // complete type in MultiBinnedConstraint test
+
+//#include <TestStatistics/LikelihoodGradientJob.h>
+#include <TestStatistics/RooUnbinnedL.h>
+#include <TestStatistics/RooBinnedL.h>
+#include <TestStatistics/RooSumL.h>
+#include <TestStatistics/optional_parameter_types.h>
+#include <TestStatistics/likelihood_builders.h>
+//#include <RooFit/MultiProcess/JobManager.h>
+//#include <RooFit/MultiProcess/ProcessManager.h> // need to complete type for debugging
```
Include or don't include

hageboeck commented on roofit/roofitcore/test/TestStatistics/testLikelihoodSerial.cxx:
```diff
@@ -0,0 +1,444 @@
+   auto nll0 = nll->getVal();
+
+   likelihood =
+//      std::make_shared<RooFit::TestStatistics::RooBinnedL>(pdf, data, RooFit::GlobalObservables(*w.var("alpha_bkg_obs")));
+//      std::make_shared<RooFit::TestStatistics::RooUnbinnedL>(pdf, data);
+//      std::make_shared<RooFit::TestStatistics::RooSumL>(pdf, data, RooFit::TestStatistics::GlobalObservables({*w.var("alpha_bkg_obs")}));
```
delete

hageboeck commented on roofit/roofitcore/test/TestStatistics/testLikelihoodSerial.cxx:
```diff
@@ -0,0 +1,444 @@
+   w.factory("ExtendPdf::egB(Gaussian::gB(x,mB[-2,-10,10],s),nB[100])") ;
+   w.factory("SIMUL::model(index[A,B],A=gA,B=gB)") ;
+
+   RooDataSet* dA = w.pdf("gA")->generate(*w.var("x"),1) ;
```
leak

hageboeck commented on roofit/roofitcore/test/TestStatistics/testLikelihoodSerial.cxx:
```diff
@@ -0,0 +1,444 @@
+
+      // Generate template histograms
+
+      RooDataHist* h_sigA = w.pdf("gA")->generateBinned(*w.var("x"),1000) ;
```
leak.
Test should be leak free if possible, because then they can be used for leak checking the rest of RooFit.

egpbos replied:
Done, also lots of leaks in RooRealL.cpp, fixed them.

hageboeck commented on roofit/roofitcore/test/TestStatistics/testLikelihoodSerial.cxx:
```diff
@@ -0,0 +1,444 @@
+   EXPECT_EQ(nll0, nll2);
+   EXPECT_DOUBLE_EQ(nll1, nll2);
+
+//   printf("nll0: %a\tnll1: %a\tnll2: %a\tnll_ts.getResult(): %a\tnll_ts.offset: %a\tnll_ts.offset_carry: %a\tlikelihood.get_carry: %a\tcarry1: %a\n", nll0, nll1, nll2, nll_ts.getResult(), nll_ts.offset(), nll_ts.offsetCarry(), likelihood->get_carry(), carry1);
```
delete

hageboeck reviewed (changes requested):
Hello!

I understand it's a big thing, but unfortunately it's far from ready. Please excuse that the comments become shorter and shorter, but I literally went through 4,008 additions.

I now also realise that there is a very nice description of what's going on in the commit message, which should be findable in doxygen. I recommend to copy those paragraphs to a findable location in doxygen. The class documentation of the base classes would be one option or the (not yet existing) docstrings for the `TestStatistics` namespace.

A really big problem while maintaining RooFit was the lack of documentation in many critical places. I will hold you to the higher standard that I tried to establish, since that will help the future maintainers a lot. 🙂 

Furthermore, there's some things that should be done "in the new way", i.e.
- Almost no `TODO`s, `FIXME`s, `xxx HACK`
- No TIterator-style loops, because the range-based loops are now 30% faster, less leaks and easier to read
- I don't accept code that's commented out and destined to rot, as nobody is compiling and checking it.
- Includes from within ROOT should be in `" "`.
- And the Kahan summation should be unified as you already anticipated.
- There's code that has been modernised / sped up while this feature here was in development, and it seems that one should look at both worlds, and take the best of both. Committing it like this would set us back two years, losing possible improvements, e.g. due to vectorisation.

For hot loops, one should consider to either make them auto-vectorisable by e.g. using ROOT's KahanSum (but this requires that users compile ROOT with vectorisation support) or (better) to put them in Manos' mini computation library that gets compiled for multiple architecture. This will get the best speed on every platform. That's of course much more work, though ... 🙁

egpbos commented:
@hageboeck thanks for another Herculean reviewing effort :) Your comments make total sense; indeed, all the copy-pasted stuff still has to be merged with all the modernization and optimization work that was done in the past two-years.

I will go through as much of your suggestions as I can before I will be on leave after tomorrow for three weeks.

After that, if anyone else has time to work on some of the issues, I'd of course welcome the help. As you know, my time on the project is running out, so I probably won't have time to integrate everything. For instance, Manos' mini computation library I have only learned about in the last few months and don't know it in enough detail to make any kind of sensible estimate of how to do it, let alone how much time this would take.

Let's discuss detailed plans in our meeting at 14:00 today.

egpbos commented on roofit/roofitcore/inc/RooMinimizer.h:
```diff
@@ -150,4 +167,50 @@ class RooMinimizer : public TObject {
+
+// static function
+template <typename LikelihoodWrapperT, typename LikelihoodGradientWrapperT>
+std::unique_ptr<RooMinimizer> RooMinimizer::create(std::shared_ptr<RooFit::TestStatistics::RooAbsL> likelihood) {
+   return std::unique_ptr<RooMinimizer>(new RooMinimizer(likelihood, static_cast<LikelihoodWrapperT*>(nullptr),
+                                                         static_cast<LikelihoodGradientWrapperT*>(nullptr)));
+}
```
Changed to `make_unique`.

egpbos commented on roofit/roofitcore/inc/TestStatistics/RooRealL.h:
```diff
@@ -0,0 +1,63 @@
+   inline TObject *clone(const char *newname) const override { return new RooRealL(*this, newname); }
+
+   // virtual overrides:
+   inline double globalNormalization() const
```
It isn't actually. Removed.

egpbos commented on roofit/roofitcore/inc/TestStatistics/RooRealL.h:
```diff
@@ -0,0 +1,63 @@
+   inline Double_t defaultErrorLevel() const override { return 0.5; }
+
+private:
+   std::shared_ptr<RooAbsL> likelihood_;
+   mutable double eval_carry = 0;
+   RooSetProxy vars_proxy_; // sets up client-server connections
```
Hm yeah, at some point I started putting the underscores after the names, because apparently the C++ standard reserves use of underscore-prefixed names. I'm aware this breaks RooFit convention. Did you ever discuss this point? I'm fine with reverting to prefix, of course, but just making sure you're aware.

egpbos commented on roofit/roofitcore/inc/TestStatistics/RooRealL.h:
```diff
@@ -0,0 +1,63 @@
+   mutable double eval_carry = 0;
+   RooSetProxy vars_proxy_; // sets up client-server connections
+
+   ClassDefOverride(RooRealL, 0);
```
Good question. Maybe I/O should be possible, since this is a RooAbsReal. Its intended purpose is to also give the new likelihood classes an old-style interface, for instance for plotting or just plugging the number into something that expects a RooAbsReal (*other than minimization, of course*, which should just be done with the new classes). Do you think there could be situations where this requires I/O? And also: what requirements would this put on the other new classes that don't have any I/O yet either?

egpbos commented on roofit/roofitcore/inc/TestStatistics/RooSubsidiaryL.h:
```diff
@@ -0,0 +1,65 @@
+   void constOptimizeTestStatistic(RooAbsArg::ConstOpCode opcode, bool doAlsoTrackingOpt) override;
+
+private:
+   std::string parent_pdf_name_;
+   RooArgList subsidiary_pdfs_{"subsidiary_pdfs"}; // Set of subsidiary PDF or "constraint" terms
+   RooArgSet parameter_set_{"parameter_set"};      // Set of parameters to which constraints apply
```
See https://github.com/root-project/root/pull/8700/files#r673851551

egpbos commented on roofit/roofitcore/inc/TestStatistics/RooAbsL.h:
```diff
@@ -0,0 +1,129 @@
+#include "RooAbsArg.h" // enum ConstOpCode
+
+#include <cstddef> // std::size_t
+#include <string>
```
I will add it to the big review commit.

egpbos commented on roofit/roofitcore/inc/TestStatistics/RooRealL.h:
```diff
@@ -0,0 +1,63 @@
+#ifndef ROOT_ROOFIT_TESTSTATISTICS_RooRealL
+#define ROOT_ROOFIT_TESTSTATISTICS_RooRealL
+
+#include <RooAbsReal.h>
```
Added to review commit.

egpbos commented:
I went through all comments, closed the ones I fixed and left some further comments for follow up discussion. The fixes I made so far are summarized in the commit message. The obvious big things still are `KahanSum` and adding lots of documentation. I will do these (and the other remaining todo's in the first post above this PR, https://github.com/root-project/root/pull/8700#issue-692681999) after my holidays.

egpbos commented:
During our meeting just now we decided that the highest priority issues for me to work on the coming days are:

1. Documentation
2. The top-level likelihood builder interface / PDF analysis (together with @wverkerke).

If time permits, after that I will try to harmonize the Kahan sum usages and the other remaining issues, but we may need to postpone these to another PR.

egpbos commented on roofit/roofitcore/inc/RooMinimizer.h:
```diff
@@ -150,4 +167,50 @@ class RooMinimizer : public TObject {
 } ;
 
+// include here to avoid circular dependency issues in class definitions
+#include "TestStatistics/MinuitFcnGrad.h"
```
As mentioned in the comment, it's a circular dependency. `RooMinimizer`'s template constructor needs a complete `MinuitFcnGrad` type and `MinuitFcnGrad`'s template constructor needs a complete `RooMinimizer` type.

Other options to solve this would be:
1. Split the headers into two: one file with only the class definition and one file with template and inline functions. Not so nice, imho.
2. Put all classes into one header (really bad idea).
3. Tear out the `RooMinimizer` dependency in `MinuitFcnGrad`'s constructor. Strictly speaking, it doesn't need the `RooMinimizer`, it uses it to get to the `Fitter` (which contains the `FitConfig` (which contains the parameter settings object that it *actually* needs)). I could just pass the thing it actually needs to the constructor. The slight downside of this is that it is important that the correct set of parameters is passed, and getting them directly from the `RooMinimizer`, as done now, ensures this. If the ctor caller has to pass this correct set in manually, there's a risk this will be done wrong some day.

I went with option 3 now, since actually the ctor is not the only point where the correct set of parameters has to be passed in, Synchronize has the same "issue".

egpbos commented:
I just noticed a compile error for the RooRealL.cxx test file that I introduced in the modernization commit (the second in this PR). I did an interactive rebase to fix this. Also included the RooMinimizer / MinuitFcnGrad circular dependency fix in that modernization/clean-up commit.

egpbos commented:
In dd32e6d, I factored out the unbinned RooNLLVar calculation parts (computeScalar and computeBatched) into static functions inside RooNLLVar. These are now also called from RooUnbinnedL.

RooUnbinnedL can now also do batched computation. I also added a test for this in testLikelihoodSerial. However, the test is failing, because it does not give the bit-wise exact equal answer as the old serial code. We should discuss whether this is expected and how to test for it then.

egpbos commented:
Remaining points and questions arising from @hageboeck's review:

1. https://github.com/root-project/root/pull/8700#discussion_r672508683 Should RooRealL have I/O? It is currently disabled by setting ClassDef version to 0.
2. https://github.com/root-project/root/pull/8700#discussion_r672509759 RooFit naming convention puts underscores in front of member names. This is discouraged, since underscore prefix is reserved for C++ implementations. I introduced some new members with underscore suffix instead, which is also common practice e.g. in the Google C++ guidelines. I think this is not something we should spend too much time on now, but maybe good to keep in mind for the future.
3. https://github.com/root-project/root/pull/8700#discussion_r672519457 The configuration parameters (GlobalObservables etc.): should we unify them or keep like this?
4. https://github.com/root-project/root/pull/8700#discussion_r672538377 This should be checked by Wouter.

In addition, some other remaining questions:
1. `testRooRealL.getValRooAddition` fails because it doesn't know `RooFormulaVar`... I'm not sure what this means, but I vaguely remember that (I think) @hageboeck and @cburgard were talking about this, so I was hoping you may be able to help me out with fixing that test.
2. In an ideal world I would have liked to add a serial implementation of the `LikelihoodGradientWrapper` class. Probably, we can reuse a lot of `RooGradMinimizerFcn` for this and it should be rather straightforward. However, this will take time. The advantage would be that we could add a test to this PR that covers all that is added, basically by just running a minimization. Note that I do have such a full-coverage test, but only for the MultiProcess minimization, where I implemented `LikelihoodGradientWrapper` in the parallel `LikelihoodGradientJob` class. So this test will come in "PR 7" (see overview). So, up for discussion, two choices: A: add (probably mostly useless except for testing this PR) "LikelihoodGradientSerial" class (based on RooGradMinimizerFcn) so we can add an integration test in this PR; B: wait for the integration test in PR 7.
3. Does everybody like the RooNLLVar statics?
4. Offset in the new TestStatistics classes can easily include the RooSubsidiaryL term as well, unlike with RooAbsTestStatistic, where the RooConstraintSum was a separate class that would not be included in the offset. I have now implemented a legacy mode to get bit-wise equal results, but is the new "full" mode a good idea at all? Does it make sense?

egpbos commented:
Brief conclusions based on the discussion via Zoom we just had with @lmoneta @guitargeek @manolismih and @wverkerke:

Conclusions from discussion via Zoom:

1. RooRealL does not need I/O. Basic principle: pdfs and datasets are persistable, test statistics are not.
2. Keep how it is.
3. Keep them and add docstring.
4. Removed TODO, it's fine. Added assert will make sure.

1. Jonas will investigate. Will write in PR.
2. Next PR, time constraints.
3. Yes. Good. Rename to ...Func and pass evalData as reference, not unique_ptr.
4. It's a step in the right direction. Keep like it is. Keep legacy in default to ease transition.

Additional todo's:
- Batched test: compare to batched RooNLLVar.
- Fix CI issues.
- Fix remaining failing tests.

egpbos commented:
The latest force push just now implements some of the points we discussed.

One thing I omitted is the change of the `evalData` parameter from a `unique_ptr` to a reference. We discussed that a pointer means that the object may be changed and this is indeed what can happen. The function, in fact, initializes the pointer if it is null. I don't know if this was intended or maybe it should be changed, but it was already like this, so I'm hesitant to change it.

Remaining todos:
- [ ] Add doxygen for the `GlobalObservables` etc parameters.
- [ ] Fix CI issues.
- [ ] Follow @guitargeek's coming comment about how to fix the `RooFormulaVar` bug in testRooRealL.

egpbos commented:
I the latest force push, I tried to fix all remaining compile errors and warnings that the CI brought up. I used fixup commits with interactive rebase to keep the commit history clean and actually make sure the earlier commits compile as well now.

Evidently, I have failed to fix the esoteric error with ambiguous overload in the strongly typed optional parameter classes. There is actually even a paper about this (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1008r0.pdf), which I thought should provide a solution, but apparently not on all compilers, so I'll have to try something else.

egpbos commented:
@phsft-bot build just on windows10/cxx14

egpbos commented:
@phsft-bot build

egpbos commented:
The latest force push should fix build configuration on Windows, which was the only remaining failing build. The cause seems to have been the extra `m` in `LIBRARIES` in the `ROOT_ADD_GTEST` statement for `testLikelihoodSerial`.

egpbos commented:
@phsft-bot build

egpbos commented:
@phsft-bot build just on windows10/cxx14

egpbos commented:
@phsft-bot build just on windows10/default

egpbos commented:
Fixed the segfault in the RooRealL tests above, now only the `RooFormulaVar` failures remaining.

guitargeek commented on roofit/roofitcore/test/TestStatistics/RooRealL.cpp:
```diff
@@ -0,0 +1,334 @@
+   if (dynamic_cast<RooAddition *>(nll) != nullptr) {
+      std::cout << "the NLL object is a RooAddition*..." << std::endl;
+      bool has_rooconstraintsum = false;
+      for (const auto nll_component : *nll->getComponents()) {
```
```suggestion
      for (const auto nll_component : static_cast<RooAddition *>(nll)->list()) {
```
This should be the correct way to fix the unit test. We don't want to loop over all the branches of the model, which is what `getComponents()` does. Instead, we want to iterate over the terms in the `RooAddition`.

egpbos replied:
This works, thanks! I did a fixup/rebase to the "modernize and clean up RooFit::TestStatistics" commit (abcea9a) to keep the commit history clean.

egpbos commented:
Disabled in Windows a test in RooRealL that doesn't work under Windows.

egpbos commented:
All remaining CI issues are not caused by this PR.

If someone wants to review again, feel free. If not, I think we can merge this now.

Note that I tried to keep commit history clean and used fixup commits + rebase to also make sure all commits compile and pass tests so it should be possible to merge without squashing. Since I made a few changes in the calculation area (Kahan sums, batch computation) which could affect precision and hence results (and hence tests) it would be nice if the commit history of this PR could be included fully.

lmoneta commented:
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

lmoneta commented:
Thank you Patrick for the update. I am testing it now again with all tests and tutorials

guitargeek commented on roofit/roofitcore/src/RooNLLVar.cxx:
```diff
@@ -459,44 +459,52 @@ Double_t RooNLLVar::evaluatePartition(std::size_t firstEvent, std::size_t lastEv
+}
+
+// static function, also used from TestStatistics::RooUnbinnedL
+RooNLLVar::ComputeResult RooNLLVar::computeBatchedFunc(const RooAbsPdf *pdfClone, RooAbsData *dataClone,
```
```suggestion
RooNLLVar::ComputeResult RooNLLVar::computeBatchedFunc(const RooAbsPdf & pdf, RooAbsData & data,
```
It's not relevant that these are `*Clone` at this point, and it's better to use references and not pointers if they can't be null.

guitargeek commented on roofit/roofitcore/src/RooNLLVar.cxx:
```diff
@@ -567,21 +575,28 @@ RooNLLVar::ComputeResult RooNLLVar::computeBatched(std::size_t stepSize, std::si
+}
 
+// static function, also used from TestStatistics::RooUnbinnedL
+RooNLLVar::ComputeResult RooNLLVar::computeScalarFunc(const RooAbsPdf *pdfClone, RooAbsData *dataClone,
```
```suggestion
RooNLLVar::ComputeResult RooNLLVar::computeScalarFunc(const RooAbsPdf & pdf, RooAbsData const& data,
```

guitargeek commented on roofit/roofitcore/inc/RooDataHist.h:
```diff
@@ -222,6 +227,8 @@ class RooDataHist : public RooAbsData, public RooDirItem {
   friend class RooAbsCachedPdf ;
   friend class RooAbsCachedReal ;
   friend class RooDataHistSliceIter ;
+  // for access into copied dataset:
+  friend class RooFit::TestStatistics::RooAbsL;
```
The changes in this file are not necessary, `RooAbsL` is not using any non-public functions in RooDataHist.

egpbos replied:
Ah, weird. I thought this was necessary for RooAbsL::initClones, but maybe I've edited the dependency out.

guitargeek commented on roofit/roofitcore/inc/TestStatistics/MinuitFcnGrad.h:
```diff
@@ -0,0 +1,166 @@
+
+// static function
+template <typename LikelihoodWrapperT, typename LikelihoodGradientWrapperT>
+MinuitFcnGrad *
+MinuitFcnGrad::create(const std::shared_ptr<RooFit::TestStatistics::RooAbsL> &likelihood, RooMinimizer *context,
+                      std::vector<ROOT::Fit::ParameterSettings> &parameters, bool verbose)
+{
```
```suggestion
std::unique_ptr<MinuitFcnGrad>
MinuitFcnGrad::create(const std::shared_ptr<RooFit::TestStatistics::RooAbsL> &likelihood, RooMinimizer *context,
                      std::vector<ROOT::Fit::ParameterSettings> &parameters, bool verbose)
{
```

Can we return a smart pointer instead?

guitargeek reviewed (approved):
Thanks Patrick, this PR looks good as it is!

* new test statistic classes are introduced to RooFit
* review comment by @hageboeck were addressed
* the things I don't like myself will be addressed by me in a follow up PR so @egpbos doesn't have to spent time on this he doesn't have
* I don't agree that the test failures on Windows are necessarily unrelated to this PR, because the failures are related to RooFit. It seems the RooFit library is unexpectedly loaded now, because the RooFit banner appears in PyROOT even if RooFit is not used. But given that it's difficult to debug windows failures, this can be investigated after merging this PR (and maybe it's a good excuse to get rid of the banner).

guitargeek commented on roofit/roofitcore/inc/TestStatistics/ConstantTermsOptimizer.h:
```diff
@@ -0,0 +1,40 @@
+ *   WV, Wouter Verkerke, UC Santa Barbara, verkerke@slac.stanford.edu       *
+ *   DK, David Kirkby,    UC Irvine,         dkirkby@uci.edu                 *
+ *                                                                           *
+ * Copyright (c) 2000-2021, Regents of the University of California          *
+ *                          and Stanford University. All rights reserved.    *
```
I'll do this, about to open a PR with some follow-ups to the multiprocessing
